### ISSUE TITLE
errcheck is called incorrectly and silently fails

### ISSUE THREAD

**Issue by horgh** [alecthomas/gometalinter#318]
I typically call `gometalinter` via `vim-go` which runs a command like this:

```
gometalinter --disable-all --enable=vet --enable=golint --enable=errcheck \
/home/vagrant/go/src/github.com/horgh/test
```

I noticed today after updating `gometalinter` that errors it normally would catch were not being reported. It appears there was a change to how it calls `errcheck` and the above command now fails the `errcheck` call.

If I revert to an earlier `gometalinter` commit, linting with `errcheck` works again.

I've updated using `gometalinter --install` as well.

I used `git bisect` and found the first commit with an issue is `849d41c4e4702571db3c9b8c646a7ff21d48bdc8`.

Here is debug output of the problem case (`b4ebfc554d8f36bfef1f180ad0aaaaac99b430d5`):

```
vagrant@ct100-test:~/go/src/github.com/horgh/test$ gometalinter --debug --disable-all --enable=vet --enable=golint --enable=errcheck /home/vagrant/go/src/github.com/horgh/test
DEBUG: PATH=/home/vagrant/bin:/home/vagrant/bin/ratpoison-git/src:/home/vagrant/go/bin:/home/vagrant/bin:/home/vagrant/bin/ratpoison-git/src:/home/vagrant/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/home/vagrant/.fzf/bin
DEBUG: GOPATH=/home/vagrant/go
DEBUG: GOBIN=
DEBUG: linting path /home/vagrant/go/src/github.com/horgh/test
DEBUG: executing /usr/bin/go tool vet /home/vagrant/go/src/github.com/horgh/test/test.go
DEBUG: executing /home/vagrant/go/bin/golint -min_confidence 0.800000 /home/vagrant/go/src/github.com/horgh/test
DEBUG: golint hits 0: ^(?P<path>.*?\.go):(?P<line>\d+):(?P<col>\d+):\s*(?P<message>.*)$
DEBUG: golint linter took 11.488242ms
DEBUG: executing /home/vagrant/go/bin/errcheck -abspath /home/vagrant/go/src/github.com/horgh/test
DEBUG: warning: /home/vagrant/go/bin/errcheck returned exit status 2: import "/home/vagrant/go/src/github.com/horgh/test": cannot import absolute path
error: failed to check packages: could not type check: no initial packages were loaded

DEBUG: errcheck hits 0: ^(?P<path>.*?\.go):(?P<line>\d+):(?P<col>\d+):\s*(?P<message>.*)$
DEBUG: errcheck linter took 9.964403ms
DEBUG: vet hits 0: ^(?:vet:.*?\.go:\s+(?P<path>.*?\.go):(?P<line>\d+):(?P<col>\d+):\s*(?P<message>.*))|(?:(?P<path>.*?\.go):(?P<line>\d+):\s*(?P<message>.*))$
DEBUG: vet linter took 50.874178ms
DEBUG: total elapsed time 57.944781ms
```

Here is debug output of the working case (`1bcfd85`):

```
vagrant@ct100-test:~/go/src/github.com/horgh/test$ gometalinter --debug --disable-all --enable=vet --enable=golint --enable=errcheck /home/vagrant/go/src/github.com/horgh/test
DEBUG: PATH=/home/vagrant/bin:/home/vagrant/bin/ratpoison-git/src:/home/vagrant/go/bin:/home/vagrant/bin:/home/vagrant/bin/ratpoison-git/src:/home/vagrant/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/home/vagrant/.fzf/bin
DEBUG: GOPATH=/home/vagrant/go
DEBUG: GOBIN=
DEBUG: linting path /home/vagrant/go/src/github.com/horgh/test
DEBUG: linting with golint: golint -min_confidence {min_confidence} {path} (on /home/vagrant/go/src/github.com/horgh/test)
DEBUG: executing /home/vagrant/go/bin/golint ["-min_confidence" "0.800000" "."]
DEBUG: linting with errcheck: errcheck -abspath {path} (on /home/vagrant/go/src/github.com/horgh/test)
DEBUG: executing /home/vagrant/go/bin/errcheck ["-abspath" "."]
DEBUG: golint hits 0: ^(?P<path>.*?\.go):(?P<line>\d+):(?P<col>\d+):\s*(?P<message>.*)$
DEBUG: golint linter took 18.766998ms
DEBUG: linting with vet: go tool vet {path}/*.go (on /home/vagrant/go/src/github.com/horgh/test)
DEBUG: executing /usr/bin/go ["tool" "vet" "test.go"]
DEBUG: vet hits 0: ^(?:vet:.*?\.go:\s+(?P<path>.*?\.go):(?P<line>\d+):(?P<col>\d+):\s*(?P<message>.*))|(?:(?P<path>.*?\.go):(?P<line>\d+):\s*(?P<message>.*))$
DEBUG: vet linter took 153.805345ms
DEBUG: warning: errcheck -abspath . returned exit status 2
DEBUG: errcheck hits 2: ^(?P<path>.*?\.go):(?P<line>\d+):(?P<col>\d+):\s*(?P<message>.*)$
DEBUG: errcheck linter took 656.594344ms
DEBUG: nolint: parsing /home/vagrant/go/src/github.com/horgh/test/test.go for directives
DEBUG: nolint: parsing /home/vagrant/go/src/github.com/horgh/test/test.go took 617.432µs
/home/vagrant/go/src/github.com/horgh/test/test.go:19:6:warning: error return value not checked (cannot convert 5 (untyped int constant) to string) (errcheck)
/home/vagrant/go/src/github.com/horgh/test/test.go:18:2:warning: error return value not checked (a declared but not used) (errcheck)
DEBUG: total elapsed time 672.846156ms
```

The silent failure seems related to #310. But I don't see an open issue for the change in the call to `errcheck`.

Thanks for your assistance!

**Comment by alecthomas** [comment]
@dnephin ping

**Comment by dnephin** [comment]
I've been away for the last few days. I should be able to fix this
tomorrow. I think the problem is that some linters expect package paths,
not file paths to directories. Relative paths work for both cases. It looks
like if the command line arg is an absolute path we don't convert it to a
relative path.

Does it work with a relative path?

I don't think my warn on panic PR will handle this case

**Comment by dnephin** [comment]
I think I have a fix in https://github.com/alecthomas/gometalinter/compare/master...dnephin:fix-errcheck-partition if you want to try it out. I need to write some tests to verify it, and see if any other linters need this fix.

I'm still curious if it works if you use a relative path (I believe it does).

If no other linters require this change it might be easier to fix errcheck instead of continuing to support another type of linter.

**Comment by alecthomas** [comment]
Yeah that seems pretty annoying TBH.

**Comment by dnephin** [comment]
Ok, it looks like lots of linters actually expect packages, so I'll finish writing the test for this and open a PR. The full list is:

* aligncheck
* errcheck
* gosimple
* interfacer
* megacheck
* safesql
* staticcheck
* structcheck
* test
* testify
* unconvert
* unparam
* unused
* varcheck

**Comment by horgh** [comment]
Works great!

Thanks so much for fixing this!

**Comment by Megalepozy** [comment]
Hey,

Seems like this error (at least in regard to errcheck) is back.

When running `gometalinter $absoulteFileNameWithPath$` errcheck is not being implemented (or ignored?) while running it with $relativeFileNameWithPath$ does run it.

Sorry if this is the wrong place to submit this issue, was wondering if I should open a new issue or not :)
